### PR TITLE
Docker module log-opt support

### PR DESCRIPTION
**Pull request by daviddyball**
- Added support for Docker engine `log-opt` arguments
- Fixed possible issue whereby the wrong method was called for (`get`/`create`)`_host_config`. Otherwise the options for `log_opt` were never propagated to the containers.

### CODE REVIEW

File: cloud/docker/docker.py [hunk 1]
```diff
@@ -109,6 +109,13 @@
       - none
       - syslog
     version_added: "2.0"
+  log_opt:
+    description:
+      - Additional options to pass to the logging driver selected above. See Docker log-driver
+        documentation for more information (https://docs.docker.com/reference/logging/overview/).
+        Requires docker >=1.7.0.
+    required: false
+    default: null
```
Comment (abadger) [inline cloud/docker/docker.py:118]: need a version_added: "2.0"

File: cloud/docker/docker.py [hunk 2]
```diff
@@ -317,6 +324,7 @@
     - "Joshua Conner (@joshuaconner)"
     - "Pavel Antonov (@softzilla)"
     - "Ash Wilson (@smashwilson)"
+    - "David Dyball"
```
Comment (abadger) [inline cloud/docker/docker.py]: If you want to be a maintainer that's fine.  Add your github id as well so we know who to ping :-)

File: cloud/docker/docker.py [hunk 3]
```diff
@@ -1418,11 +1453,11 @@ def create_containers(self, count=1):
                   'stdin_open':   self.module.params.get('stdin_open'),
                   'tty':          self.module.params.get('tty'),
                   'cpuset':       self.module.params.get('cpu_set'),
-                  'host_config':  self.create_host_config(),
+                  'host_config':  self.get_host_config(),
                   'user':         self.module.params.get('docker_user'),
                   }
         if self.ensure_capability('host_config', fail=False):
-            params['host_config'] = self.get_host_config()
+            params['host_config'] = self.create_host_config()
```
Comment (abadger) [inline cloud/docker/docker.py:1428]: This is a little suspect. I'll explain more in the pull request itself.

**Comment by abadger** [comment]
Few comments on the PR.  Those should be easy to fix, though.

The one problem i see is something that seems to be an existing problem that your PR is highlighting.  That's what is happening with get_host_config() vs create_host_config().

After some research into the pull requests that lead to both of those:
- create_host_config: https://github.com/ansible/ansible-modules-core/pull/1405
- get_host_config: https://github.com/ansible/ansible-modules-core/pull/1208 and https://github.com/ansible/ansible-modules-core/pull/1744

I think we had two PRs that implemented very similar changes to fix the same issue.  Because the names were different we ended up merging both of them when we should have realized they were attempting to do the same thing and only merged one.

I'll try to straighten out the code right now to only keep one of those methods.  If you could rebase and test after I push that I'd appreciate it.

**Comment by abadger** [comment]
here's a pull request to merge those two functions: https://github.com/ansible/ansible-modules-core/pull/2011

I'm going to see if any community people using docker will respond to it overnight and then merge it tomorrow.

**Comment by abadger** [comment]
@daviddyball okay, merged #2011 so you should be able to update this PR to only add log_opt to the new create_host_config() and it should work.

**Comment by daviddyball** [comment]
Thanks @abadger. Sorry for the delay in picking this back up. I'll take a look at it now.

**Comment by daviddyball** [comment]
@abadger re-based on latest devel after #2011 merge and addressed comments. Not 100% I would make a good maintainer, so took myself off the list ;-)

**Comment by abadger** [comment]
@daviddyball Thanks for adding this!

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!

**Pull request merged**